### PR TITLE
[Core] Refactoring sampler and support prompt logprob for chunked prefill 

### PR DESCRIPTION
**Pull request by rkooo567**
Summary;

**Refactoring sampling metadata and sampler. More concretely**
- Introduce SequenceGroupToSample class instead of having multiple list data structure to combine
- Instead of writing index looping logic again and again, we prepare the prefill/decode indices ahead of time and reuse it
- Move prepare_sample to SamplingMetadata
- Remove all indexing logics that assume requests are entire prefill or decode
- Introduce `do_sample` to `SequenceGroupMetadata`. If it is set to False, sampling/sample logprob calculation for the corresponding seq_group is skipped.
- Improve docstring and confusing variable names 
- Remove `perform_sampling` because it is leaky and overlaps with `do_sample`. I just use `is_driver_worker` directly for the same purpose.

**Fix prompt logprob for chunked prefil**
- Existing logic has strong assumption where the seq group only contains entire prefill or decode.
- Allow to incrementally update prompt logprobs
- Allow to skip sampling when it is chunked prefill (since it is not required)

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -459,9 +459,12 @@ def has_unfinished_requests(self) -> bool:
...
 
     def _process_model_outputs(
-            self, output: List[SamplerOutput],
-            scheduled_seq_groups: List[SequenceGroup],
-            ignored_seq_groups: List[SequenceGroup]) -> List[RequestOutput]:
+        self,
+        output: List[SamplerOutput],
+        scheduled_seq_groups: List[SequenceGroup],
+        ignored_seq_groups: List[SequenceGroup],
+        seq_group_metadata_list: List[SequenceGroupMetadata],
```
Comment (rkooo567) [inline vllm/engine/llm_engine.py:466]: Decided to keep it because removing it introduces a little ugly code

File: vllm/engine/output_processor/multi_step.py [hunk 2]
```diff
@@ -44,6 +44,12 @@ def __init__(
         self.get_tokenizer_for_seq = get_tokenizer_for_seq
         self.stop_checker = stop_checker
 
+    def process_prompt_logprob(self, seq_group: SequenceGroup,
+                               outputs: List[SequenceGroupOutput]) -> None:
+        # TODO(sang): Prompt logprob currently not implemented in multi step
```
Comment (rkooo567) [inline vllm/engine/output_processor/multi_step.py:49]: cc @cadedaniel

File: vllm/engine/output_processor/single_step.py [hunk 3]
```diff
@@ -55,17 +55,23 @@ def process_outputs(self, sequence_group: SequenceGroup,
...
-    def _process_sequence_group_outputs(self, seq_group: SequenceGroup,
-                                        outputs: SequenceGroupOutput) -> None:
-
-        # Process prompt logprobs
-        prompt_logprobs = outputs.prompt_logprobs
-        if prompt_logprobs is not None and \
-            seq_group.sampling_params.detokenize and self.detokenizer:
+    def process_prompt_logprob(self, seq_group: SequenceGroup,
+                               outputs: List[SequenceGroupOutput]) -> None:
+        assert len(outputs) == 1, ("Single step should only has 1 output.")
```
Comment (rkooo567) [inline vllm/engine/output_processor/single_step.py:60]: cc @cadedaniel is this assumption correct?

File: vllm/worker/cpu_model_runner.py [hunk 4]
```diff
@@ -252,98 +254,101 @@ def _prepare_decode(
...
-        sampling_metadata = SamplingMetadata(
-            seq_groups=seq_groups,
-            seq_data=seq_data,
-            prompt_lens=prompt_lens,
-            selected_token_indices=selected_token_indices,
-            categorized_sample_indices=categorized_sample_indices,
-            generators=generators,
-        )
-        return sampling_metadata
+    # def _prepare_sample(
```
Comment (rkooo567) [inline vllm/worker/cpu_model_runner.py:257]: Will remove after confirming tests are passed

File: vllm/worker/neuron_model_runner.py [hunk 5]
```diff
@@ -141,105 +141,105 @@ def _prepare_decode(
...
-        sampling_metadata = SamplingMetadata(
-            seq_groups=seq_groups,
-            seq_data=seq_data,
-            prompt_lens=prompt_lens,
-            selected_token_indices=selected_token_indices,
-            categorized_sample_indices=categorized_sample_indices,
-            generators=generators,
-        )
-        return sampling_metadata
+    # def _prepare_sample(
```
Comment (rkooo567) [inline vllm/worker/neuron_model_runner.py:144]: Will remove after confirming tests are passed

File: vllm/model_executor/sampling_metadata.py [hunk 6]
```diff
@@ -6,57 +6,267 @@
...
+        else:
+            # Decode
+            prompt_logprob_len = 0
+            sample_len = len(seq_ids) if do_sample else 0
+
+        # Update indices to select from the model output.
+        """
+        hidden_states = model(...)
+        logits = hidden_states[selected_token_indices]
+        """
```
Comment (zhuohan123) [inline vllm/model_executor/sampling_metadata.py:222]: What is this comment for?

File: vllm/engine/output_processor/multi_step.py (hunk 2 above)
Comment (zhuohan123) [reply vllm/engine/output_processor/multi_step.py:49]: Print a warning here?

**Review by zhuohan123 — APPROVED** [review]
Thank you for the refactor SangBin! The code looks much better now. I left some small comments. But in general the code looks pretty good to me.

File: vllm/model_executor/sampling_metadata.py (hunk 6 above)
Comment (rkooo567) [reply vllm/model_executor/sampling_metadata.py:222]: Oh just wanted to show how this data structure is used later. Let me add additional comments.

File: vllm/model_executor/layers/sampler.py [hunk 7]
```diff
@@ -429,13 +482,11 @@ def _sample_with_torch(
...
         if num_tokens == 0:
             continue
-        seq_group_ids = categorized_seq_group_ids[sampling_type]
-        seq_groups = [sampling_metadata.seq_groups[i] for i in seq_group_ids]
-        is_prompts = [i < sampling_metadata.num_prompts for i in seq_group_ids]
-        sample_metadata[sampling_type] = (seq_group_ids, seq_groups,
-                                          is_prompts, sample_indices)
-        long_sample_indices = sample_indices.long()
 
+        seq_group_idx = categorized_seq_group_idx[sampling_type]
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply vllm/model_executor/layers/sampler.py:486]: reverted to id. I thought it was confusing because each seq_group already has request_id (and it doesn't match). But no strong preference.

**Comment by rkooo567** [comment]
@zhuohan123 thanks for the review! All comments are addressed!

File: vllm/engine/output_processor/single_step.py (hunk 3 above)
Comment (cadedaniel) [reply vllm/engine/output_processor/single_step.py:60]: correct!
